In [1]:
import pandas as pd

In [2]:
df = pd.read_csv('https://raw.githubusercontent.com/datasciencedojo/datasets/master/titanic.csv')

In [3]:
df

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500   NaN        S  
..     ...               ...      ...   ...      ...  
886      0            211536  13.0000   NaN        S  
887      0            112053  30.0000   B42        S  
888      2        W./C. 6607  23.4500   NaN        S  
889      0            111369  30.0000  C148        C  
890      0            370376   7.7500   NaN        Q  

[891 rows x 12 columns]

In [4]:
df.dtypes

PassengerId      int64
Survived         int64
Pclass           int64
Name            object
Sex             object
Age            float64
SibSp            int64
Parch            int64
Ticket          object
Fare           float64
Cabin           object
Embarked        object
dtype: object

In [5]:
df.describe() # works only for numerical data  columns

PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   14.526497    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   20.125000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   38.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  891.000000  891.000000  
mean     0.381594   32.204208  
std      0.806057   49.693429  
min      0.000000    0.000000  
25%      0.000000    7.910400  
50%      0.000000   14.454200  
75%      0.000000   31.000000  
max      6.000000  512.329200

1. 25% - 25% of data is smaller than this value
1. 50% - 50% of data is smaller than this value
1. 75% - 75% of data is smaller than this value
1. max - max value of data 
1. min - min value of data
1. count - total values in that column ( example in every row 891 but in age only 714 values are present)
1. mean - average
1. std :- standard deviation - how much individual data points in a dataset differ from the mean

In [6]:
# selecting by data type
df.dtypes=='object'

PassengerId    False
Survived       False
Pclass         False
Name            True
Sex             True
Age            False
SibSp          False
Parch          False
Ticket          True
Fare           False
Cabin           True
Embarked        True
dtype: bool

In [8]:
type(df.dtypes=='object') # this is a series with index as column name

pandas.core.series.Series

In [9]:
df.dtypes[df.dtypes == 'object'] # get the data from the series which is going to follow this condition

Name        object
Sex         object
Ticket      object
Cabin       object
Embarked    object
dtype: object

In [14]:
df.dtypes[df.dtypes == 'object'].index

Index(['Name', 'Sex', 'Ticket', 'Cabin', 'Embarked'], dtype='object')

In [15]:
df[df.dtypes[df.dtypes == 'object'].index].describe() # get the data which follows this condition - descibes for categorical data

Name   Sex  Ticket Cabin Embarked
count                   891   891     891   204      889
unique                  891     2     681   147        3
top     Dooley, Mr. Patrick  male  347082    G6        S
freq                      1   577       7     4      644

unique - unique data
count -  in how many cells data is present
top :- data on top cell
freq :- frequency of top coming in it

In [17]:
df.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [20]:
df.dtypes=='float64'

PassengerId    False
Survived       False
Pclass         False
Name           False
Sex            False
Age             True
SibSp          False
Parch          False
Ticket         False
Fare            True
Cabin          False
Embarked       False
dtype: bool

In [25]:
df[df.dtypes[df.dtypes=='float64'].index]

Age     Fare
0    22.0   7.2500
1    38.0  71.2833
2    26.0   7.9250
3    35.0  53.1000
4    35.0   8.0500
..    ...      ...
886  27.0  13.0000
887  19.0  30.0000
888   NaN  23.4500
889  26.0  30.0000
890  32.0   7.7500

[891 rows x 2 columns]

In [26]:
df.columns

Index(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp',
       'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked'],
      dtype='object')

In [29]:
df['Ticket'][4:11] # performing fn likes list

4      373450
5      330877
6       17463
7      349909
8      347742
9      237736
10    PP 9549
Name: Ticket, dtype: object

In [30]:
# making a new column and saving data in it
df['new_col']= 0 # if already exists then overwrite

In [31]:
df.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  new_col  
0      0         A/5 21171   7.2500   NaN        S        0  
1      0          PC 17599  71.2833   C85        C        0  
2      0  STON/O2. 3101282   7.9250   NaN        S        0  
3      0            113803  53.1000  C123        S        0  
4      0            373450   8.0500   NaN        S        0

In [33]:
df['p_id_plus_class']= df['PassengerId']+df['Pclass']

In [34]:
df.head(10)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   
5            6         0       3   
6            7         0       1   
7            8         0       3   
8            9         1       3   
9           10         1       2   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   
5                                   Moran, Mr. James    male   NaN      0   
6                            McCarthy, Mr. Timothy J    male  54.0      0   
7                     Palsson, Master. Gosta Leonard    male   2.0      3   
8  Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.0      0   
9                Nasser, Mrs. Nicholas (Adele Achem)  female  14.0      1   

   Parch            Ticket     Fare Cabin Embarked  new_col  p_id_plus_class  
0      0         A/5 21171   7.2500   NaN        S        0                4  
1      0          PC 17599  71.2833   C85        C        0                3  
2      0  STON/O2. 3101282   7.9250   NaN        S        0                6  
3      0            113803  53.1000  C123        S        0                5  
4      0            373450   8.0500   NaN        S        0                8  
5      0            330877   8.4583   NaN        Q        0                9  
6      0             17463  51.8625   E46        S        0                8  
7      1            349909  21.0750   NaN        S        0               11  
8      2            347742  11.1333   NaN        S        0               12  
9      0            237736  30.0708   NaN        C        0               12

In [37]:
pd.Categorical(df['Pclass']) # shows unique category of particular column here 3 values example in this code is 3

[3, 1, 3, 1, 3, ..., 2, 1, 3, 1, 3]
Length: 891
Categories (3, int64): [1, 2, 3]

In [40]:
pd.Categorical(df['Sex'])

['male', 'female', 'female', 'female', 'male', ..., 'male', 'female', 'female', 'male', 'male']
Length: 891
Categories (2, object): ['female', 'male']

In [41]:
df['Cabin'].unique() # shows all the unique values including nan - not a number

array([nan, 'C85', 'C123', 'E46', 'G6', 'C103', 'D56', 'A6',
       'C23 C25 C27', 'B78', 'D33', 'B30', 'C52', 'B28', 'C83', 'F33',
       'F G73', 'E31', 'A5', 'D10 D12', 'D26', 'C110', 'B58 B60', 'E101',
       'F E69', 'D47', 'B86', 'F2', 'C2', 'E33', 'B19', 'A7', 'C49', 'F4',
       'A32', 'B4', 'B80', 'A31', 'D36', 'D15', 'C93', 'C78', 'D35',
       'C87', 'B77', 'E67', 'B94', 'C125', 'C99', 'C118', 'D7', 'A19',
       'B49', 'D', 'C22 C26', 'C106', 'C65', 'E36', 'C54',
       'B57 B59 B63 B66', 'C7', 'E34', 'C32', 'B18', 'C124', 'C91', 'E40',
       'T', 'C128', 'D37', 'B35', 'E50', 'C82', 'B96 B98', 'E10', 'E44',
       'A34', 'C104', 'C111', 'C92', 'E38', 'D21', 'E12', 'E63', 'A14',
       'B37', 'C30', 'D20', 'B79', 'E25', 'D46', 'B73', 'C95', 'B38',
       'B39', 'B22', 'C86', 'C70', 'A16', 'C101', 'C68', 'A10', 'E68',
       'B41', 'A20', 'D19', 'D50', 'D9', 'A23', 'B50', 'A26', 'D48',
       'E58', 'C126', 'B71', 'B51 B53 B55', 'D49', 'B5', 'B20', 'F G63',
       'C62 C64',

In [49]:
# filtering all the data set where age is less than 18
df['Age']<18

0      False
1      False
2      False
3      False
4      False
       ...  
886    False
887    False
888    False
889    False
890    False
Name: Age, Length: 891, dtype: bool

In [52]:
df[df['Age']<18]

PassengerId  Survived  Pclass                                     Name  \
7              8         0       3           Palsson, Master. Gosta Leonard   
9             10         1       2      Nasser, Mrs. Nicholas (Adele Achem)   
10            11         1       3          Sandstrom, Miss. Marguerite Rut   
14            15         0       3     Vestrom, Miss. Hulda Amanda Adolfina   
16            17         0       3                     Rice, Master. Eugene   
..           ...       ...     ...                                      ...   
850          851         0       3  Andersson, Master. Sigvard Harald Elias   
852          853         0       3                  Boulos, Miss. Nourelain   
853          854         1       1                Lines, Miss. Mary Conover   
869          870         1       3          Johnson, Master. Harold Theodor   
875          876         1       3         Najib, Miss. Adele Kiamie "Jane"   

        Sex   Age  SibSp  Parch    Ticket     Fare Cabin Embarked  new_col  \
7      male   2.0      3      1    349909  21.0750   NaN        S        0   
9    female  14.0      1      0    237736  30.0708   NaN        C        0   
10   female   4.0      1      1   PP 9549  16.7000    G6        S        0   
14   female  14.0      0      0    350406   7.8542   NaN        S        0   
16     male   2.0      4      1    382652  29.1250   NaN        Q        0   
..      ...   ...    ...    ...       ...      ...   ...      ...      ...   
850    male   4.0      4      2    347082  31.2750   NaN        S        0   
852  female   9.0      1      1      2678  15.2458   NaN        C        0   
853  female  16.0      0      1  PC 17592  39.4000   D28        S        0   
869    male   4.0      1      1    347742  11.1333   NaN        S        0   
875  female  15.0      0      0      2667   7.2250   NaN        C        0   

     p_id_plus_class  
7                 11  
9                 12  
10                14  
14                18  
16                20  
..               ...  
850              854  
852              856  
853              855  
869              873  
875              879  

[113 rows x 14 columns]

In [56]:
len(df[df['Age']<18])

113

1. df.describe().loc['mean','Fare'] # method 1 : select mean row from Fare column
2. mean_fare = df.describe().at['mean', 'Fare'] # method 2. Using .at[] (For a Single Value)
3. mean_fare = df.describe().T['Fare']['mean'] # method 3 : Using .transpose() + .mean (Alternative) This works because .T (transpose) flips rows and columns, making "Fare" a row index.


In [79]:
df[df['Fare'] > float(df.describe().loc['mean','Fare'])] # who paid more Fare than average


PassengerId  Survived  Pclass  \
1              2         1       1   
3              4         1       1   
6              7         0       1   
23            24         1       1   
27            28         0       1   
..           ...       ...     ...   
856          857         1       1   
863          864         0       3   
867          868         0       1   
871          872         1       1   
879          880         1       1   

                                                  Name     Sex   Age  SibSp  \
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
6                              McCarthy, Mr. Timothy J    male  54.0      0   
23                        Sloper, Mr. William Thompson    male  28.0      0   
27                      Fortune, Mr. Charles Alexander    male  19.0      3   
..                                                 ...     ...   ...    ...   
856         Wick, Mrs. George Dennick (Mary Hitchcock)  female  45.0      1   
863                  Sage, Miss. Dorothy Edith "Dolly"  female   NaN      8   
867               Roebling, Mr. Washington Augustus II    male  31.0      0   
871   Beckwith, Mrs. Richard Leonard (Sallie Monypeny)  female  47.0      1   
879      Potter, Mrs. Thomas Jr (Lily Alexenia Wilson)  female  56.0      0   

     Parch    Ticket      Fare        Cabin Embarked  new_col  p_id_plus_class  
1        0  PC 17599   71.2833          C85        C        0                3  
3        0    113803   53.1000         C123        S        0                5  
6        0     17463   51.8625          E46        S        0                8  
23       0    113788   35.5000           A6        S        0               25  
27       2     19950  263.0000  C23 C25 C27        S        0               29  
..     ...       ...       ...          ...      ...      ...              ...  
856      1     36928  164.8667          NaN        S        0              858  
863      2  CA. 2343   69.5500          NaN        S        0              867  
867      0  PC 17590   50.4958          A24        S        0              869  
871      1     11751   52.5542          D35        S        0              873  
879      1     11767   83.1583          C50        C        0              881  

[211 rows x 14 columns]

In [83]:
# use multiple condition using operators
df[(df['Fare'] == 0) & (df['Sex'] == 'male')]


PassengerId  Survived  Pclass                              Name   Sex  \
179          180         0       3               Leonard, Mr. Lionel  male   
263          264         0       1             Harrison, Mr. William  male   
271          272         1       3      Tornquist, Mr. William Henry  male   
277          278         0       2       Parkes, Mr. Francis "Frank"  male   
302          303         0       3   Johnson, Mr. William Cahoone Jr  male   
413          414         0       2    Cunningham, Mr. Alfred Fleming  male   
466          467         0       2             Campbell, Mr. William  male   
481          482         0       2  Frost, Mr. Anthony Wood "Archie"  male   
597          598         0       3               Johnson, Mr. Alfred  male   
633          634         0       1     Parr, Mr. William Henry Marsh  male   
674          675         0       2        Watson, Mr. Ennis Hastings  male   
732          733         0       2              Knight, Mr. Robert J  male   
806          807         0       1            Andrews, Mr. Thomas Jr  male   
815          816         0       1                  Fry, Mr. Richard  male   
822          823         0       1   Reuchlin, Jonkheer. John George  male   

      Age  SibSp  Parch  Ticket  Fare Cabin Embarked  new_col  p_id_plus_class  
179  36.0      0      0    LINE   0.0   NaN        S        0              183  
263  40.0      0      0  112059   0.0   B94        S        0              265  
271  25.0      0      0    LINE   0.0   NaN        S        0              275  
277   NaN      0      0  239853   0.0   NaN        S        0              280  
302  19.0      0      0    LINE   0.0   NaN        S        0              306  
413   NaN      0      0  239853   0.0   NaN        S        0              416  
466   NaN      0      0  239853   0.0   NaN        S        0              469  
481   NaN      0      0  239854   0.0   NaN        S        0              484  
597  49.0      0      0    LINE   0.0   NaN        S        0              601  
633   NaN      0      0  112052   0.0   NaN        S        0              635  
674   NaN      0      0  239856   0.0   NaN        S        0              677  
732   NaN      0      0  239855   0.0   NaN        S        0              735  
806  39.0      0      0  112050   0.0   A36        S        0              808  
815   NaN      0      0  112058   0.0  B102        S        0              817  
822  38.0      0      0   19972   0.0   NaN        S        0              824

In [90]:
df[(df['Fare'] == 0) & (df['Survived'] == 1)]

PassengerId  Survived  Pclass                          Name   Sex   Age  \
271          272         1       3  Tornquist, Mr. William Henry  male  25.0   

     SibSp  Parch Ticket  Fare Cabin Embarked  new_col  p_id_plus_class  
271      0      0   LINE   0.0   NaN        S        0              275

In [91]:
max(df['Fare'])

512.3292

In [93]:
df[df['Fare']==max(df['Fare'])] # who paid maximun fare

PassengerId  Survived  Pclass                                Name  \
258          259         1       1                    Ward, Miss. Anna   
679          680         1       1  Cardeza, Mr. Thomas Drake Martinez   
737          738         1       1              Lesurer, Mr. Gustave J   

        Sex   Age  SibSp  Parch    Ticket      Fare        Cabin Embarked  \
258  female  35.0      0      0  PC 17755  512.3292          NaN        C   
679    male  36.0      0      1  PC 17755  512.3292  B51 B53 B55        C   
737    male  35.0      0      0  PC 17755  512.3292         B101        C   

     new_col  p_id_plus_class  
258        0              260  
679        0              681  
737        0              739

In [94]:
df.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  new_col  p_id_plus_class  
0      0         A/5 21171   7.2500   NaN        S        0                4  
1      0          PC 17599  71.2833   C85        C        0                3  
2      0  STON/O2. 3101282   7.9250   NaN        S        0                6  
3      0            113803  53.1000  C123        S        0                5  
4      0            373450   8.0500   NaN        S        0                8

In [96]:
df[0:2] # get 0th and 1st row

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   

   Parch     Ticket     Fare Cabin Embarked  new_col  p_id_plus_class  
0      0  A/5 21171   7.2500   NaN        S        0                4  
1      0   PC 17599  71.2833   C85        C        0                3

In [99]:
df[['Fare','Name','Sex']][0:2]

Fare                                               Name     Sex
0   7.2500                            Braund, Mr. Owen Harris    male
1  71.2833  Cumings, Mrs. John Bradley (Florence Briggs Th...  female

The main difference between .loc and .iloc in Pandas is that .loc is used for label-based indexing, allowing you to select rows and columns by their names, while .iloc is used for integer-based indexing, allowing you to select rows and columns by their numerical positions. This means .loc includes the endpoint in a range, whereas .iloc does not.

In [101]:
df.loc[0:2,['PassengerId','Fare']] # get 1st 2 rows from Columns Passenger Id and Fare

PassengerId     Fare
0            1   7.2500
1            2  71.2833
2            3   7.9250

In [106]:
df.loc[0:2,['PassengerId','Fare']] # 1st param is always row and 2nd is Fare

PassengerId     Fare
0            1   7.2500
1            2  71.2833
2            3   7.9250

In [107]:
df.iloc[0:2,[0,1]] # iloc requires the index for columns too

PassengerId  Survived
0            1         0
1            2         1

#### using custom list as index
custom_index = ['A100', 'A101', 'A102', 'A103', 'A104']  # Example custom index list

df = pd.read_csv('data.csv')

df.index = custom_index  # Set the custom list as index

print(df.head())
